### PR TITLE
fix(mempool): Pass cancellable context to iterator

### LINKED ISSUE #4057: mempool: Iterator cannot be cancelled
The following test is sometimes failing. This is because the mempool iterator is blocked waiting for new transactions and there's no way to cancel its execution.
```
--- FAIL: TestBroadcastTxForPeerStopsWhenPeerStops (10.06s)
    leaktest.go:132: leaktest: timed out checking goroutines
    leaktest.go:[15](https://github.com/cometbft/cometbft/actions/runs/10790390214/job/29925500859?pr=4040#step:6:16)0: leaktest: leaked goroutine: goroutine 2903827 [chan receive]:
        github.com/cometbft/cometbft/mempool.(*CListIterator).WaitNextCh.func1()
        	/home/runner/work/cometbft/cometbft/mempool/clist_mempool.go:861 +0x12c
        created by github.com/cometbft/cometbft/mempool.(*CListIterator).WaitNextCh in goroutine 2901968
        	/home/runner/work/cometbft/cometbft/mempool/clist_mempool.go:857 +0x13d
    leaktest.go:150: leaktest: leaked goroutine: goroutine 5614247 [runnable]:
        time.when(0x14)
        	/opt/hostedtoolcache/go/1.22.7/x64/src/time/sleep.go:28 +0x7e
        time.(*Timer).Reset(0xc000348410, 0x14)
        	/opt/hostedtoolcache/go/1.22.7/x64/src/time/sleep.go:138 +0x4f
        github.com/cometbft/cometbft/internal/timer.(*ThrottleTimer).fireRoutine(0xc0003d6980)
        	/home/runner/work/cometbft/cometbft/internal/timer/throttle_timer.go:46 +0x1e5
        created by time.goFunc
        	/opt/hostedtoolcache/go/1.22.7/x64/src/time/sleep.go:[17](https://github.com/cometbft/cometbft/actions/runs/10790390214/job/29925500859?pr=4040#step:6:18)7 +0x45
```

### PR DESCRIPTION
**Pull request by hvanz**
Fixes #4057

---

#### PR checklist

- [ ] Tests written/updated
- [ ] Changelog entry added in `.changelog` (we use [unclog](https://github.com/informalsystems/unclog) to manage our changelog)
- [ ] Updated relevant documentation (`docs/` or `spec/`) and code comments

### CODE REVIEW

**Review by jmalicevic — APPROVED** [review]

File: mempool/reactor.go [hunk 1]
```diff
@@ -217,24 +217,29 @@ func (memR *Reactor) broadcastTxRoutine(peer p2p.Peer) {
...
-	iter := memR.mempool.NewIterator()
-	var entry Entry
+	ctx, cancel := context.WithCancel(context.Background())
+	go func() {
+		select {
+		case <-peer.Quit():
+			cancel()
+		case <-memR.Quit():
+			cancel()
+		case <-ctx.Done():
```
Comment (melekes) [inline mempool/reactor.go:227]: This is never reached.

File: mempool/clist_mempool.go [hunk 2]
```diff
@@ -850,21 +852,25 @@ func (iter *CListIterator) WaitNextCh() <-chan Entry {
 		if iter.cursor == nil {
 			// We are at the beginning of the iteration or the saved entry got removed: wait until
 			// the list becomes not empty and select the first entry.
-			<-iter.txs.WaitChan()
+			select {
+			case <-iter.txs.WaitChan():
+			case <-iter.ctx.Done():
```
Comment (melekes) [inline mempool/clist_mempool.go:857]: Do you think we shouldn't exit here?

```suggestion
			case <-iter.ctx.Done():
			   close(ch)
			   return
```

File: mempool/clist_mempool.go [hunk 3]
```diff
@@ -850,21 +852,25 @@ func (iter *CListIterator) WaitNextCh() <-chan Entry {
...
+			}
 			// Note that Front can return nil.
 			iter.cursor = iter.txs.Front()
 		} else {
 			// Wait for the next entry after the current one.
-			<-iter.cursor.NextWaitChan()
+			select {
+			case <-iter.cursor.NextWaitChan():
+			case <-iter.ctx.Done():
```
Comment (melekes) [inline mempool/clist_mempool.go:865]: Do you think we shouldn't exit here?

```suggestion
			case <-iter.ctx.Done():
			   close(ch)
			   return
```

**Review by melekes — CHANGES_REQUESTED** [review]

File: mempool/reactor.go (hunk 1 above)
Comment (hvanz) [reply mempool/reactor.go]: That's correct. I removed it.

**Review by melekes — APPROVED** [review]

**Pull request merged**